src/lib/tools.ts

```diff
@@ -0,0 +1,314 @@
+}): Promise<string> {
+	// TODO: support git URI
+	if (isValidGitUri(uri)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathUri(uri)) {
+		return (await fs.readFile(join(uri, path))).toString('utf8');
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+/**
+ * Lists a path in a URI
+ * @param {string} URI
+ * @param {string} path
+ * @param {string} cachePath
+ * @returns {Promise<string[]>}
+ */
+export async function listUri({
+	uri,
+	path,
+	cachePath,
+}: {
+	uri: string;
+	path: string;
+	cachePath?: string;
+}): Promise<string[]> {
+	// TODO: support git URI
+	if (isValidGitUri(uri)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathUri(uri)) {
+		return await fs.readdir(join(uri, path));
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+/**
+ * Keyframe unwrapper to a standard format
+ * @param productRepoURI: URI of the product repo
+ * @param keyFramePath: path of keyframe
+ *
+ * returns: Keyframe object
+ */
+export async function unwrapKeyframe(
+	productRepoURI: string,
+	keyFramePath: string = './keyframe.yml',
+): Promise<object> {
+	// TODO: keyframe layering
+	let keyFrame = await loadFromUri({ uri: productRepoURI, path: keyFramePath });
+
+	if (keyFrame) {
+		keyFrame = _.filter(
+			_.get(keyFrame, ['children', 'sw', 'containerized-application'], []),
+			component => component.type === 'sw.containerized-application',
+		);
+		keyFrame = _.mapValues(_.keyBy(keyFrame, 'slug'), o =>
+			_.merge(_.get(o, 'assets', {}), { version: _.get(o, 'version') }),
+		);
+		return keyFrame;
+	} else {
+		return {};
+	}
+}
+
+/**
+ * Creates an ssh tunnel for executing a promise
+ * @param tnlConfig: ssh2 tunnel configuration object
+ * @param prom: promise
+ * @param timeout: tunnel timeout.
+ * @returns {Promise<any>}
+ */
+export async function runInTunnel(
+	tnlConfig: tunnel.Config,
+	prom: Promise<any>,
+	timeout: number,
+): Promise<any> {
+	const tnl = await tunnelAsync(tnlConfig);
+	const wait = setTimeout(function() {
+		tnl.close();
+		throw new TimeoutError('Timeout exceeded');
+	}, timeout);
+	return prom.then((ret: any) => {
+		clearTimeout(wait);
+		if (tnl) {
+			tnl.close();
+		}
+		return ret;
+	});
+}
+
+/**
+ * Convert a nested configMap object to a flat Key-Value pair configMap
+ * @param {ConfigMap} configMap
+ * @returns {ConfigMap}
+ */
+export function configMapToPairs(configMap: ConfigMap): ConfigMap {
+	const keyPaths: string[] = [];
+	const ret: ConfigMap = {};
+	function traverse(configMap: any, path: string = '') {
+		if (configMap && _.isObject(configMap)) {
+			_.forIn(configMap, function(value: any, key: string) {
+				traverse(value, `${path}.${key}`);
+			});
+		} else {
+			keyPaths.push(_.trimStart(path, '.'));
+		}
+	}
+	traverse(configMap);
+	for (const keyPath of keyPaths) {
+		ret[_.replace(keyPath, new RegExp('\\.', 'g'), '___')] = _.get(
```

Comment:
I am really not sure; @mikesimos @hedss maybe you could shed some light?

Reply:
I'm guessing it's because it's trying to convert things to work with the lodash `.` path notation, eg `_.get(a, 'b.c.d')` where you cannot use a `.` within the path, but that'd be easily solved by using `_.get(a, ['b', 'c.d'])` style if so